### ISSUE TITLE
Got internal error when trying to use ScopeIf node

### ISSUE THREAD

**Issue by yuvalrakavy** [DynamoDS/Dynamo#9111]
If this issue is with Dynamo for Revit, please post your issue on the [Dynamo for Revit Issues page](https://github.com/DynamoDS/DynamoRevit/issues).

If this issue is **not** a bug report or improvement request, please check the [Dynamo forum](https://forum.dynamobim.com/), and start a thread there to discuss your issue.

## Dynamo version

(Which version of Dynamo are you using? Go to Help > About if you're not sure.)
1.3.0.946

## Operating system
Windows 10
(e.g. Windows 7, Windows 8.1, etc)

## What did you do? 
I switched from If node (which worked incorrectly) to ScopeIf node (which should have solved my problem) and I go an internal error

(Fill in here)

## What did you expect to see?
The case if that I have a list of curves. In one case I need to use the whole list, and on the other case I need use the list while dropping the last item. 

 If I use If node to select between the cases, I always get the list with the last item dropped
 If I try to use ScopeIf node, I get the internal error message.

In a small case of using a list of numbers, If node always returned the list with last item dropped, and scopeIf node worked correctly (returning either the full list, or the list without the last item)

(Fill in here)

## What did you see instead?
I will split my custom node into two separate nodes for handling the two cases (avoiding the If). This is ugly as the user will have to pick which one to use rather having the node work for both cases (the cases are if the input curve is closed or not)

The internal error I got is:

Exception of type 'ProtoCore.Exceptions.CompilerInternalException' was thrown.

   at ProtoFFI.CLRObjectMarshler.UnMarshal(StackValue dsObject, Context context, Interpreter dsi, Type expectedCLRType)
   at ProtoFFI.CLRFFIFunctionPointer.Execute(Context c, Interpreter dsi)
   at ProtoCore.Lang.FFIFunctionEndPoint.Execute(Context c, List`1 formalParameters, StackFrame stackFrame, RuntimeCore runtimeCore)
   at ProtoCore.CallSite.ExecWithZeroRI(List`1 functionEndPoint, Context c, List`1 formalParameters, StackFrame stackFrame, RuntimeCore runtimeCore, FunctionGroup funcGroup, SingleRunTraceData previousTraceData, SingleRunTraceData newTraceData)
   at ProtoCore.CallSite.Execute(List`1 functionEndPoint, Context c, List`1 formalParameters, List`1 replicationInstructions, StackFrame stackFrame, RuntimeCore runtimeCore, FunctionGroup funcGroup)
   at ProtoCore.CallSite.DispatchNew(Context context, List`1 arguments, List`1 partialReplicationGuides, DominantListStructure domintListStructure, StackFrame stackFrame, RuntimeCore runtimeCore)
   at ProtoCore.DSASM.Executive.Callr(Int32 blockDeclId, Int32 functionIndex, Int32 classIndex, Boolean& explicitCall, Boolean isDynamicCall, Boolean hasDebugInfo)
   at ProtoCore.DSASM.Executive.CALLR_Handler(Instruction instruction)
   at ProtoCore.DSASM.Executive.Execute(Int32 exeblock, Int32 entry, Language language)
   at ProtoCore.DSASM.Executive.Execute(Int32 exeblock, Int32 entry, List`1 breakpoints, Language language)
   at ProtoCore.DSASM.Interpreter.Run(Int32 codeblock, Int32 entry, Language lang, List`1 breakpoints)
   at ProtoCore.Lang.JILFunctionEndPoint.Execute(Context c, List`1 formalParameters, StackFrame stackFrame, RuntimeCore runtimeCore)
   at ProtoCore.CallSite.ExecWithZeroRI(List`1 functionEndPoint, Context c, List`1 formalParameters, StackFrame stackFrame, RuntimeCore runtimeCore, FunctionGroup funcGroup, SingleRunTraceData previousTraceData, SingleRunTraceData newTraceData)
   at ProtoCore.CallSite.ExecWithRISlowPath(List`1 functionEndPoint, Context c, List`1 formalParameters, List`1 replicationInstructions, StackFrame stackFrame, RuntimeCore runtimeCore, FunctionGroup funcGroup, SingleRunTraceData previousTraceData, SingleRunTraceData newTraceData)
   at ProtoCore.CallSite.ExecWithRISlowPath(List`1 functionEndPoint, Context c, List`1 formalParameters, List`1 replicationInstructions, StackFrame stackFrame, RuntimeCore runtimeCore, FunctionGroup funcGroup, SingleRunTraceData previousTraceData, SingleRunTraceData newTraceData)
   at ProtoCore.CallSite.Execute(List`1 functionEndPoint, Context c, List`1 formalParameters, List`1 replicationInstructions, StackFrame stackFrame, RuntimeCore runtimeCore, FunctionGroup funcGroup)
   at ProtoCore.CallSite.DispatchNew(Context context, List`1 arguments, List`1 partialReplicationGuides, DominantListStructure domintListStructure, StackFrame stackFrame, RuntimeCore runtimeCore)
   at ProtoCore.DSASM.Executive.Callr(Int32 blockDeclId, Int32 functionIndex, Int32 classIndex, Boolean& explicitCall, Boolean isDynamicCall, Boolean hasDebugInfo)
   at ProtoCore.DSASM.Executive.CALLR_Handler(Instruction instruction)
   at ProtoCore.DSASM.Executive.Execute(Int32 exeblock, Int32 entry, Language language)
   at ProtoCore.DSASM.Executive.Execute(Int32 exeblock, Int32 entry, List`1 breakpoints, Language language)
   at ProtoCore.DSASM.Executive.BounceUsingExecutive(Executive executive, Int32 exeblock, Int32 entry, StackFrame stackFrame, Int32 locals, Boolean fepRun, Executive exec, List`1 breakpoints)
   at ProtoScript.Runners.ProtoScriptRunner.ExecuteLive(Core core, RuntimeCore runtimeCore)
   at ProtoScript.Runners.LiveRunner.Execute(Boolean isCodeCompiled)
   at ProtoScript.Runners.LiveRunner.CompileAndExecute(List`1 astList)
   at ProtoScript.Runners.LiveRunner.CompileAndExecuteForDeltaExecution(List`1 astList)
   at ProtoScript.Runners.LiveRunner.SynchronizeInternal(GraphSyncData syncData)
   at ProtoScript.Runners.LiveRunner.UpdateGraph(GraphSyncData syncData)
   at Dynamo.Scheduler.UpdateGraphAsyncTask.HandleTaskExecutionCore()
   at Dynamo.Scheduler.AsyncTask.Execute()
(Fill in here)

**Comment by sameyers** [comment]
I had a similar error when testing was meant to return results on unequal list length.  Dynamo v1.3, Revit 2017; here is the error log:

Runtime exception occured. The Type of StackValue is not Double

   at ProtoCore.DSASM.StackValue.get_DoubleValue()
   at ProtoCore.DSASM.Executive.IsNodeModified(StackValue svGraphNode, StackValue svUpdateNode)
   at ProtoCore.DSASM.Executive.XLangUpdateDependencyGraph(Int32 currentLangBlock)
   at ProtoCore.DSASM.Executive.RestoreFromBounce()
   at ProtoCore.DSASM.Executive.RETB_Handler()
   at ProtoCore.DSASM.Executive.Execute(Int32 exeblock, Int32 entry, Language language)
   at ProtoCore.DSASM.Executive.Execute(Int32 exeblock, Int32 entry, List`1 breakpoints, Language language)
   at ProtoCore.DSASM.Executive.BounceUsingExecutive(Executive executive, Int32 exeblock, Int32 entry, StackFrame stackFrame, Int32 locals, Boolean fepRun, Executive exec, List`1 breakpoints)
   at ProtoScript.Runners.ProtoScriptRunner.ExecuteLive(Core core, RuntimeCore runtimeCore)
   at ProtoScript.Runners.LiveRunner.Execute(Boolean isCodeCompiled)
   at ProtoScript.Runners.LiveRunner.CompileAndExecute(List`1 astList)
   at ProtoScript.Runners.LiveRunner.CompileAndExecuteForDeltaExecution(List`1 astList)
   at ProtoScript.Runners.LiveRunner.SynchronizeInternal(GraphSyncData syncData)
   at ProtoScript.Runners.LiveRunner.UpdateGraph(GraphSyncData syncData)
   at Dynamo.Scheduler.UpdateGraphAsyncTask.HandleTaskExecutionCore()
   at Dynamo.Scheduler.AsyncTask.Execute()

**Comment by johnpierson** [comment]
@yuvalrakavy , are you able to provide some sample files to reproduce on our end?

Also, have you experienced this in 2.0.x versions?

**Comment by johnpierson** [comment]
Going to go ahead and close this for now as we need sample files to continue. If this is still an issue for you, please feel free to reopen and attach the files needed for us to test.

**Comment by KonradZaremba** [comment]
I have the same bug in 2.16.1
It is crushing a quite big definition with the same error. I have tried to narrow down problem. The component works on simple data like single text but when I add a list of points or more complex objects - crashes.  My attempt at isolating the problem ended up with this setup: 
![image](https://github.com/DynamoDS/Dynamo/assets/18317435/f94bbc40-b1a2-44b3-afdf-fbc37320a7cd)
I got a Cyclic Dependency warning on the Point component and Scope If.
